ipsec: fix stale keys reclaim logic

jibi opened the pull request:
The current logic used to reclaim stale IPSec keys is incorrect. In
fact, whenever a new key is added we just spawn a new goroutine which
after linux_defaults.IPsecKeyDeleteDelay time will delete all keys
different than the one just added.

The issue with this approach is that if a new key is added while one of
these goroutine is still running, this goroutine will end up deleting
the new key.

This commit fixes this by introducing a single new goroutine that
periodically removes any stale key (XFRM state and XFRM policy template)
after enough time has passed since the current key has been added.

Fixes: https://github.com/cilium/cilium/pull/19814

jibi committed:
ipsec: fix stale keys reclaim logic

The current logic used to reclaim stale IPSec keys is incorrect. In
fact, whenever a new key is added we just spawn a new goroutine which
after linux_defaults.IPsecKeyDeleteDelay time will delete all keys
different than the one just added.

The issue with this approach is that if a new key is added while one of
these goroutine is still running, this goroutine will end up deleting
the new key.

This commit fixes this by introducing a single new goroutine that
periodically removes any stale key (XFRM state and XFRM policy) after
linux_defaults.IPsecKeyDeleteDelay has passed since the key was
replaced.

Fixes: #19814
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

wip

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -710,3 +690,61 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
+
+	xfrmStateList, err := netlink.XfrmStateList(0)
+	if err != nil {
+		scopedLog.WithError(err).Warning("deleting previous SPI, xfrm state list error")
```
If I'm a user and I read this, I'll assume it deleted the previous key and then failed while listing XFRM states. Maybe best to stick to `Failed to list XFRM states`?

kaworu commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -65,10 +66,22 @@ type ipSecKey struct {
 
 var (
 	ipSecKeysGlobalLock lock.RWMutex
```
Should this mutex be renamed since after this patch it is used to protect more than `ipSecKeysGlobal` (e.g. `doReclaimStaleKeys()` write-lock it without ever accessing `ipSecKeysGlobal`)?

kaworu reviewed (approved):
Thanks @jibi patch LGTM.

jibi committed:
ipsec: rename ipSecKeysGlobalLock to ipSecLock

as now this lock is used to protect also ipSecCurrentKeySPI in addition
to ipSecKeysGlobal.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -710,3 +690,61 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
+
+		scopedLog := scopedLog.WithField("oldSPI", s.Spi)
+
+		scopedLog.Info("New encryption keys reclaiming SPI")
```
```suggestion
		scopedLog.Info("Deleting stale XFRM state.")
```

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -710,3 +690,61 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
+		scopedLog := scopedLog.WithField("oldSPI", s.Spi)
+
+		scopedLog.Info("New encryption keys reclaiming SPI")
+		if err := netlink.XfrmStateDel(&s); err != nil {
```
What about the XFRM policies?

pchaigno reviewed (changes requested):
I think we also need to clear stale XFRM policies, no?

pchaigno reviewed (changes requested):
What happens if we crash right after loading the new key? Agent will restart, not find anything in the map of stale keys, and remove all XFRM state and policies from previous SPI, right? That will cause connectivity issues with other nodes that have not started using the new SPI yet.

Instead of trying to maintain this state, why can't we rely on the creation timestamps of XFRM states and policies to know if they are older than the timeout? We have those under e.g. `xfrmState.Statistics.AddTime`.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -710,3 +700,84 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
+	nextPolicy:
+		for _, p := range xfrmPolicyList {
+			if ipSecPolicyHasTemplateForSPI(p, s.Spi) {
+				if err := xfrmDeleteTemplateFromPolicy(p, s.Spi); err != nil {
```
We want to delete the policy, not just its templates (we always have only one).

The SPI is also part of the XFRM policy's mark, though I'm not sure if it's best to check that or the template's SPI field.

jibi replied:
I've switched to the XFRM policy's mark test since it simplifies the code and we can assume there's always only one template per policy

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -260,24 +265,8 @@ func ipSecReplacePolicyOut(src, dst, tmplSrc, tmplDst *net.IPNet, dir IPSecDir)
-		}
-	}
+func ipSecPolicyForSPI(policy netlink.XfrmPolicy, spi int) bool {
+	return int(policy.Mark.Value>>12) == spi
```
We should define this value as a constant at the top. We already use it in two places without a proper constant, but this time it's going to be easier to miss if we ever change the value. The two other places will break connectivity if they have conflicting shift values; here it may just cause XFRM states to leak, which our CI won't surface.

It can probably be a separate PR if you want to avoid another iteration through CI.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -710,3 +676,83 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
+			if err := netlink.XfrmPolicyDel(&p); err != nil {
+				scopedLog.WithError(err).Warning("Deleting stale XFRM policy failed")
+			}
+		}
```
You can move this second loop outside the first one if you switch `ipSecPolicyForSPI()` to a parser of the SPI (e.g., `getSPIFromXfrmPolicy()`) and just run through the list of XFRM policies, deleting those with a stale SPI, same as you're doing for XFRM states.

pchaigno reviewed (approved):
Two minor comments below.

jibi committed:
ipsec: add ipSecXfrmMark{Set,Get}SPI helpers

which allows to encode and extract an SPI value in and from a XfrmMark

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -253,33 +265,13 @@ func ipSecReplacePolicyOut(src, dst, tmplSrc, tmplDst *net.IPNet, dir IPSecDir)
-			}
-		}
-	}
-}
```
Shouldn't this removal be in a different commit?

jibi replied:
Yes, messed up with rebase, it should be in the subsequent commit

edit: we actually still need it for tests, I'll update `doReclaimStaleKeys()` to use it rather than inlining the logic

pchaigno replied:
`git rebase --exec "make build -j$(nproc)" origin/master` for the win :wink:

jibi committed:
logfields: add SPI and OldSPI IPSec constants

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno reviewed (approved):
LGTM! (Nice commit history!)

twpayne reviewed (approved):
👍  for CLI

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -253,29 +278,32 @@ func ipSecReplacePolicyOut(src, dst, tmplSrc, tmplDst *net.IPNet, dir IPSecDir)
 	policy.Dir = netlink.XFRM_DIR_OUT
 	policy.Mark = &netlink.XfrmMark{
-		Value: ((spiWide << 12) | linux_defaults.RouteMarkEncrypt),
+		Value: ipSecXfrmMarkSetSPI(linux_defaults.RouteMarkToProxy, key.Spi),
```
Typo I guess:
```suggestion
		Value: ipSecXfrmMarkSetSPI(linux_defaults.RouteMarkEncrypt, key.Spi),
```

jibi replied:
:scream:

jibi committed:
nodediscovery: add LocalNode method

This method will be used in a subsequent commit to retrieve a copy of
the localNode object stored in the nodeDiscovery one.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

julianwiedmann commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
+	ipSecLock.Lock()
+	defer ipSecLock.Unlock()
+
+	xfrmStateList, err := netlink.XfrmStateList(0)
```
```suggestion
	xfrmStateList, err := netlink.XfrmStateList(netlink.FAMILY_ALL)
```

julianwiedmann commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
+	// Because of this, initialize currentKeyCreatedAt to the current time
+	// as we assume the current key is still being created unless we find
+	// its state.
+	currentKeyCreatedAt := time.Now()
```
This is fine for now. But I really think we should just break out when the `ipSecCurrentKeySPI` hasn't been installed yet. Avoid any doubt whether the subsequent `time.Now()` arithmetic will catch this in all corner cases.

julianwiedmann commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
+	// the AddTime of the IPSec state for the key.
+	//
+	// The creation of the IPSec state is asynchronous to loadIPSecKeys():
+	// - loadIPSecKeys() will only set ipSecCurrentKeySPI to the the SPI of
```
```suggestion
	// - loadIPSecKeys() will only set ipSecCurrentKeySPI to the SPI of
```

julianwiedmann commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -299,6 +328,34 @@ func ipsecDeleteXfrmState(ip net.IP) {
+func ipsecDeleteXfrmPoliciesNotMatchingSPI(spi uint8) {
+	scopedLog := log.WithField(logfields.SPI, spi)
+
+	xfrmPolicyList, err := netlink.XfrmPolicyList(0)
```
```suggestion
	xfrmPolicyList, err := netlink.XfrmPolicyList(netlink.FAMILY_ALL)
```

julianwiedmann commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
+	// If the current key was added less then IPSecKeyDeleteDelay time ago,
+	// don't try to reclaim any stale state or policy
+	if time.Now().Sub(currentKeyCreatedAt) < linux_defaults.IPsecKeyDeleteDelay {
+		return
```
One (obvious) caveat is that users should not cycle their keys faster than `IPsecKeyDeleteDelay` :).

jibi replied:
Correct. The initial approach was based on maintaining some state (the creation timestamp of each key) but as Paul pointed out that may cause connectivity issues in some circumstances: https://github.com/cilium/cilium/pull/19932#pullrequestreview-983002135

Maybe we can resort to use both approaches :thinking:

odinuge replied:
I'm also curios about the situation where a new key is inserted to the k8s configmap, and a new node start up immediately. The new node will then only have the new key, while it can take a few minutes until all other nodes have the new key, since this solution will rely on the kubelet updating the configmap on disk.

In such a situation, will a new node be blocked from scheduling pod networked traffic, or will it be able to schedule pods successfully, but they won't have pod network connectivity the first few minutes.. 🤔  Haven't had time to test it yet.

julianwiedmann reviewed (approved):
lgtm, just some minor comments. Don't think we need to update `Documentation/gettingstarted/encryption-ipsec.rst` either, it's all implementation details.

jibi committed:
ipsec: replace 0 with FAMILY_ALL netlink constant

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
docs: update IPSec docs

Signed-off-by: Gilberto Bertin <jibi@cilium.io>
